app/src/org/commcare/gis/MapboxLocationPickerActivity.kt

```diff
@@ -0,0 +1,275 @@
+        confirm_location_button.visibility = if (inViewMode()) {
+            View.GONE
+        } else {
+            View.VISIBLE
+        }
+        confirm_location_button.setOnClickListener {
+            // Use the map target's coordinates to make a reverse geocoding search
+            Intent().apply {
+                putExtra(FormEntryConstants.LOCATION_RESULT, GeoUtils.locationToString(viewModel.getLocation()))
+            }.also {
+                setResult(RESULT_OK, it)
+                finish()
+            }
+        }
+        cancel_button.setOnClickListener {
+            finish()
+        }
+        switch_map.setOnClickListener {
+            currentMapStyleIndex = (currentMapStyleIndex + 1) % mapStyles.size
+            val style = Style.Builder()
+                    .fromUri(mapStyles[currentMapStyleIndex])
+                    .withImage(MARKER_ICON_IMAGE_ID, ContextCompat.getDrawable(this, R.drawable.marker)!!)
+            map.setStyle(style) {
+                val loc = viewModel.getLocation()
+                updateMarker(LatLng(loc.latitude, loc.longitude, loc.altitude))
+            }
+        }
+        current_location.setOnClickListener {
+            isManualSelectedLocation = false
+        }
+    }
+
+    private fun inViewMode() = intent.getBooleanExtra(GeoPointWidget.EXTRA_VIEW_ONLY, false)
+
+    override fun onResume() {
+        super.onResume()
+        observeViewModel()
+        requestLocation()
+    }
+
+    override fun onPause() {
+        super.onPause()
+        locationController.stop()
+    }
+
+    private fun observeViewModel() {
+        viewModel.placeName.observe(this, androidx.lifecycle.Observer {
+            confirm_location_button.isEnabled = true
+            location.text = it
+        })
+    }
+
+    override fun onDestroy() {
+        map.removeOnMapClickListener(mapClickListener)
+        locationController.destroy()
+        super.onDestroy()
+    }
+
+    override fun getMapLayout(): Int {
+        return R.layout.activity_mapbox_location_picker
+    }
+
+    override fun shouldFocusUserLocationOnLoad(): Boolean {
+        return false
+    }
+
+    override fun onMapLoaded() {
+        if (!inViewMode()) {
+            map.addOnMapClickListener(mapClickListener)
+        }
+        setMapStyle()
+    }
+
+    private fun setMapStyle() {
+        val style = Style.Builder()
+                .fromUri(mapStyles[currentMapStyleIndex])
+                .withImage(MARKER_ICON_IMAGE_ID, ContextCompat.getDrawable(this, R.drawable.marker)!!)
+        map.setStyle(style) {
+            enableLocationComponent(it)
+            addMarker(it, map.cameraPosition.target)
+            initialMarkerPosition()
+        }
+    }
+
+    private fun enableLocationComponent(style: Style) {
+        val component = map.locationComponent
+        component.activateLocationComponent(LocationComponentActivationOptions.builder(
+                this, style
+        ).build())
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/814afa0addb33f0487c018513623a7ef81feedab/app/src/org/commcare/gis/MapboxLocationPickerActivity.kt#L179"></span>Call requires permission which may be rejected by user: code should explicitly check to see if permission is available (with <code>checkPermission</code>) or explicitly handle a potential <code>SecurityException</code></td>
    </tr>
  </tbody>
</table>